### PR TITLE
HIVE-23040 : Checkpointing for repl dump incremental phase

### PR DESCRIPTION
**Pull request by pkumarsinha**

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 1]
```diff
@@ -663,6 +669,322 @@ public void testMultiDBTxn() throws Throwable {
...
+    replica.load(replicatedDbName, primaryDbName)
+            .run("select * from " + replicatedDbName + ".t1")
+            .verifyResults(new String[] {})
+            .run("select * from " + replicatedDbName + ".t2")
+            .verifyResults(new String[] {});
+
+
+    //Case 1: When the last dump finished all the events and
+    //only  _finished_dump file at the hiveDumpRoot was about to be written when it failed.
+    ReplDumpWork.testDeletePreviousDumpMetaPath(true);
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:688]: Why is this needed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 2]
```diff
@@ -663,6 +669,322 @@ public void testMultiDBTxn() throws Throwable {
...
+
+    Map<String, Long> eventModTimeMap = new HashMap<>();
+    long firstIncEventID = Long.parseLong(bootstrapDump.lastReplicationId) + 1;
+    long lastIncEventID = Long.parseLong(incrementalDump1.lastReplicationId);
+    assertTrue(lastIncEventID > (firstIncEventID + 1));
+
+    for (long eventId=firstIncEventID; eventId<=lastIncEventID; eventId++) {
+      Path eventRoot = new Path(hiveDumpDir, String.valueOf(eventId));
+      if (fs.exists(eventRoot)) {
+        eventModTimeMap.put(String.valueOf(eventId), fs.getFileStatus(eventRoot).getModificationTime());
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:712]: Should check the individual files. The root level directory ts mayn't change even if contents are changing.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 3]
```diff
@@ -663,6 +669,322 @@ public void testMultiDBTxn() throws Throwable {
...
+    assertTrue(fs.getFileStatus(thirdLastEvtRoot).getModificationTime() > thirdLastEvtModTimeOld);
+
+    replica.load(replicatedDbName, primaryDbName)
+            .run("select * from " + replicatedDbName + ".t1")
+            .verifyResults(new String[] {"1", "3"})
+            .run("select * from " + replicatedDbName + ".t2")
+            .verifyResults(new String[] {"2", "4"});
+  }
+
+  @Test
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:798]: Add tests for delete/drop/add new table during checkpointing

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 4]
```diff
@@ -33,6 +33,8 @@
 import org.apache.hadoop.hive.ql.DriverFactory;
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:33]: Add tests for both distcp and regular copy

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 5]
```diff
@@ -731,9 +826,17 @@ private boolean shouldResumePreviousDump(DumpMetaData dumpMetaData) {
...
 
-  private boolean shouldResumePreviousDump(Path dumpPath) {
-    Path hiveDumpPath = new Path(dumpPath, ReplUtils.REPL_HIVE_BASE_DIR);
-    return shouldResumePreviousDump(new DumpMetaData(hiveDumpPath, conf));
+  private boolean shouldResumePreviousDump(Path lastDumpPath, Path previousValidHiveDumpPath) throws IOException {
+    if (validDump(lastDumpPath)) {
+      return false;
+    }
+    // In case of incremental _dumpmetadata may not be present, still we should resume.
+    if (previousValidHiveDumpPath != null) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:834]: We should check if a previous invalid hive dump exists. If you check for previousValidHiveDumpPath, that may be true even if previous dump path was not deleted.

Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:834]: Tests for this flow

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 6]
```diff
@@ -742,7 +845,7 @@ long currentNotificationId(Hive hiveDb) throws TException {
 
   Path dumpDbMetadata(String dbName, Path metadataRoot, long lastReplId, Hive hiveDb) throws Exception {
     // TODO : instantiating FS objects are generally costly. Refactor
-    Path dbRoot = getBootstrapDbRoot(metadataRoot, dbName, false);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:745]: This was different for bootstrap duing incremental flow
`if (isIncrementalPhase) {
      dumpRoot = new Path(dumpRoot, ReplUtils.INC_BOOTSTRAP_ROOT_DIR_NAME);
    }`
How is the path now?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java [hunk 7]
```diff
@@ -94,6 +94,12 @@
   // Configuration to enable/disable dumping ACID tables. Used only for testing and shouldn't be
   // seen in production or in case of tests other than the ones where it's required.
   public static final String REPL_DUMP_INCLUDE_ACID_TABLES = "hive.repl.dump.include.acid.tables";
+
+  // HDFS Config to define the maximum number of items a directory may contain.
+  public static final String DFS_MAX_DIR_ITEMS_CONFIG = "dfs.namenode.fs-limits.max-directory-items";
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/ReplUtils.java:99]: all other configs use . as separator instead of -

File: ql/src/java/org/apache/hadoop/hive/ql/parse/EximUtil.java [hunk 8]
```diff
@@ -76,6 +76,7 @@
   public static final String FILES_NAME = "_files";
   public static final String DATA_PATH_NAME = "data";
   public static final String METADATA_PATH_NAME = "metadata";
+  public static final String LAST_EVENT_ID_NAME = "LastEventID";
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/EximUtil.java:79]: lastEventId

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/Utils.java [hunk 9]
```diff
@@ -100,6 +105,20 @@ public Void execute() throws IOException {
     }
   }
 
+  public static boolean create(Path outputFile, HiveConf hiveConf, boolean replace)
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/Utils.java:108]: The replace param is unused

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/Utils.java [hunk 10]
```diff
@@ -64,10 +64,15 @@
 
   public static void writeOutput(List<List<String>> listValues, Path outputFile, HiveConf hiveConf)
       throws SemanticException {
+    writeOutput(listValues, outputFile, hiveConf, false);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/Utils.java:67]: Better to add FS Retry

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java (hunk 4 above)
Comment (aasha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:33]: Add tests for partial events not dumped

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 11]
```diff
@@ -663,6 +669,322 @@ public void testMultiDBTxn() throws Throwable {
...
+            .verifyResults(new String[] {"1"})
+            .run("select * from " + replicatedDbName + ".t2")
+            .verifyResults(new String[] {"100"});
+
+    dumpClause = Arrays.asList("'" + ReplUtils.DFS_MAX_DIR_ITEMS_CONFIG + "'='1000'");
+
+    WarehouseInstance.Tuple incrementalDump2 = primary.run("use " + primaryDbName)
+            .dump(primaryDbName, dumpClause);
+
+    eventCount = Integer.parseInt(incrementalDump1.lastReplicationId)
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:979]: incrementalDump1 or incrementalDump2

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 5 above)
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:834]: if previous dump path was not deleted, it would not qualify as lastDumpPath. However, have dome some refactoring to make it look more clear.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java (hunk 11 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:979]: Actually, this is redundant check.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java (hunk 2 above)
Comment (aasha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:712]: Please cross check this. Also add tests where 5 events are dumped for instance. Delete 2 events and keep 3. Only last 2 events should be rewritten

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 12]
```diff
@@ -663,6 +669,454 @@ public void testMultiDBTxn() throws Throwable {
...
+
+    eventCount = Integer.parseInt(incrementalDump2.lastReplicationId)
+            - Integer.parseInt(incrementalDump1.lastReplicationId);
+    assertTrue(eventCount > 5);
+
+    replica.load(replicatedDbName, primaryDbName)
+            .run("select * from " + replicatedDbName + ".t1")
+            .verifyResults(new String[] {"1", "2", "3", "4", "5", "6", "7", "8", "9", "10"});
+  }
+
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:1119]: Test for corner cases like deletion of the _event_id file or corrupt _event_id file. Checkpointing shouldn't be blocked on that and proceed with a new dir if resume is not possible

Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:1119]: Already have a test which covers deletion of the _event_id .

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 13]
```diff
@@ -663,6 +670,541 @@ public void testMultiDBTxn() throws Throwable {
     }
   }
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:672]: Add a test for external table incremental checkpointing

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 14]
```diff
@@ -511,6 +539,71 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
...
+    };
+    try {
+      retriable.run();
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  public void updateLastEventDumped(Path ackFile, long lastReplId) throws SemanticException {
+    List<List<String>> listValues = new ArrayList<>();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:576]: why is this a list?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 15]
```diff
@@ -425,11 +432,16 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
     EventUtils.MSClientNotificationFetcher evFetcher
         = new EventUtils.MSClientNotificationFetcher(hiveDb);
 
+
+    int maxEventLimit  = getMaxEventAllowed(work.maxEventLimit());
     EventUtils.NotificationEventIterator evIter = new EventUtils.NotificationEventIterator(
-        evFetcher, work.eventFrom, work.maxEventLimit(), evFilter);
+        evFetcher, work.eventFrom, maxEventLimit, evFilter);
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:438]: If maxEventLimit is not used from repldumpwork, it can be removed from there

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 16]
```diff
@@ -438,16 +450,24 @@ private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive
...
     replLogger.startLog();
+    long dumpedCount = resumeFrom - work.eventFrom;
+    if (dumpedCount > 0) {
+      LOG.info("Event id {} to {} are already dumped, skipping {} events", work.eventFrom, resumeFrom, dumpedCount);
+    }
+    cleanFailedEventDirIfExists(dumpRoot, resumeFrom);
     while (evIter.hasNext()) {
       NotificationEvent ev = evIter.next();
       lastReplId = ev.getEventId();
       Path evRoot = new Path(dumpRoot, String.valueOf(lastReplId));
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:465]: Path creation is not needed for the skipped events. Can be done only for the new events
